mips/atomic.c

```diff
@@ -0,0 +1,56 @@
+#include <atomic.h>
+
+#define __MIPS_PLATFORM_SYNC_NOPS ""
+static __inline void mips_sync(void) {
+  __asm __volatile(".set noreorder\n"
+                   "\tsync\n" __MIPS_PLATFORM_SYNC_NOPS ".set reorder\n"
+                   :
+                   :
+                   : "memory");
+}
+
+static __inline uint32_t atomic_cmpset_32(__volatile uint32_t *p,
```

Comment:
I copied these from FreeBSD, thus I'm pretty sure they work. I think this is also quite educational to see how to implement these functions properly, so it's not bad idea to leave them, but this isn't a good argument to keep this ;).

Anyway I don't like using internal gcc functions too much because we can never know if they are compiled or not. What if we change toolchain again and in new implementation of toolchain there won't be these operations? That's another few hours (potentially) of wasted time trying to compile/link etc. get everything to work. Also I like to have my functions documented inside code and have possibility to see implementation, if I was new to the project and saw something like `__sync_bool_compare_and_swap` I'd probably waste few minutes trying to find something on stackoverflow, while having documentation in *.h or *.c file makes everything obvious instantly, and if someone is bad at writing documentation I have implementation inside and I can check what it does. 

Also, here is a reason why FreeBSD uses their own compare and swap, right?